[MRG] Docs for deployement (#281)
paris-saclay-cds/ramp-board: doc/contribute.rst

.. _contribute:

########################
Develop and contribute
########################

Welcome to the RAMP team. We are always happy to have new RAMP developers.

> You can contribute to this code through ``Pull Request`` on GitHub_. Please, make

Reviewer: Make Pull request a link: 
```
You can contribute to this code by making a `Pull Request <https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests>`_ on Github

> sure that your code is coming with unit tests to ensure full coverage and

Reviewer: ```suggestion
sure that your code comes with unit tests to ensure full coverage and
```

continuous integration in the API.

.. _GitHub: https://github.com/paris-saclay-cds/ramp-board/pulls

Install for development
-----------------------

> To install RAMP please fallow :ref:`install guideline <install>` making sure 

Reviewer: ```suggestion
To install RAMP please follow :ref:`install guideline <install>` making sure 
```

> that you are using ``make inplace`` option.

Reviewer: ```suggestion
that you are use the ``make inplace`` option.
```

Prepare database engine
-----------------------
Next, create the database, following the steps as described here:
:ref:`create database <set_database>`

You created the ``Postgres database cluster`` using the command::

    ~ $ initdb postgres_dbs

and then you started it with::

    ~ $ pg_ctl -D postgres_dbs -l logfile start

However, RAMP tests do not know which database engine you are using for 

> your ramp development and therefore you need to let them know about it.

Reviewer: ```suggestion
your ramp development and therefore you need to inform RAMP tests about this
```

> To do it, open ``db_engine.yml`` file located in the ``ramp-board`` directory. 

Reviewer: ```suggestion
To do this, open the ``db_engine.yml`` file located in the ``ramp-board`` directory. 
```

It should look as follows::

    db_owner: postgres
    db_cluster_name: postgres

The above settings are used by ``Travis Cl`` which is a continuous integration
service hosted at ``GitHub``. Therefore, you need to change them to your local
settings.

``db_owner`` is the owner of your database. If you don't know who is 
the owner of your database you can find it out by typing in your terminal::
    
    ~ $ psql -l

> This command will list all your databases along with their owners. 

Reviewer: ```suggestion
This command will list all of your databases along with their owners. 
```

``db_cluster_name`` is the name of your Postgres database cluster (if you used

> commands as above your ``Postgres database cluster`` name is postgres_dbs).

Reviewer: ```suggestion
commands as above your ``Postgres database cluster`` name is 'postgres_dbs').
```

Test
----

> Each time before running the tests make sure your Postgres database cluster is 

Reviewer: ```suggestion
Each time before running the tests make sure your Postgres database cluster has been
```

started. You can always start it using the command::

    ~ $ pg_ctl -D postgres_dbs -l logfile start

You are now ready to run the tests. You can do so using ``pytest`` from the root ``ramp-board`` directory::

    ~ $ pytest -vsl .

The above will only work when the packages were installed in development mode.
In the other case, you can test the individual packages with::

    ~ $ pytest -vsl --pyargs ramp_utils ramp_database ramp_frontend ramp_engine

How to contribute
-----------------

This guide is adapted from `scikit-learn contribution guide`_.

.. _scikit-learn contribution guide: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md

Forking RAMP
============

The preferred way to contribute to RAMP is to fork the `ramp-board repository`_ on GitHub:

.. _ramp-board repository: https://github.com/paris-saclay-cds/ramp-board

1) To fork the `ramp-board repository`_ click on the 'Fork' button near the 
   top of the page. This creates a copy of the code under your account 
   on the GitHub server.

2) Clone this copy to your local disk::

        $ git clone git@github.com:YourLogin/ramp-board.git
        $ cd ramp-board

> 3) Create a branch to hold your changes::

Reviewer: ```suggestion
3) Create a branch (called 'my-feature' below) to hold your changes::
```

        $ git checkout -b my-feature

   and start making changes. 

.. note::
    Never work in the ``master`` branch!

4) Work on this copy on your computer using Git to do the version
   control. When you're done editing, do::

>         $ git add modified_files

Reviewer: ```suggestion
        $ git add <modified_files>
```

        $ git commit

   to record your changes in Git, then push them to GitHub with::

        $ git push -u origin my-feature

Finally, go to the web page of your fork of the ramp-board repo,
and click 'Pull request' to send your changes to the maintainers for
review. This will send an email to the committers.

If any of the above seems like magic to you, then look up `Git documentation`_ 
on the web.

.. _Git documentation: https://git-scm.com/documentation

Contributing Pull Requests
==========================

It is recommended to check that your contribution complies with the
following rules before submitting a pull request:

-  Follow the coding-guidelines_ as for scikit-learn.

-  When applicable, use the validation tools and other code in the
   `ramp_utils` subpackage.

-  If your pull request addresses an issue, please use the title to describe

>    the issue and mention the issue number in the pull request description to

Reviewer: ```suggestion
   the issue and include the issue number in the pull request description to
```

        $ pip install flake8
        $ flake8 path/to/module.py

Filing bugs
===========
We use Github issues to track all bugs and feature requests; feel free to
open an issue if you have found a bug or wish to see a feature implemented.

It is recommended to check that your issue complies with the
following rules before submitting:

-  Verify that your issue is not being currently addressed by other
   issues_ or `pull requests`_.

-  Please ensure all code snippets and error messages are formatted in
   appropriate code blocks.
   See `Creating and highlighting code blocks`_.

-  Please include your operating system type and version number, as well
   as your Python, scikit-learn, numpy, and scipy versions. This information
   can be found by runnning the following code snippet::

    python
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import numpy; print("NumPy", numpy.__version__)
    import scipy; print("SciPy", scipy.__version__)
    import sklearn; print("Scikit-Learn", sklearn.__version__)
   
-  Please be specific about what estimators and/or functions are involved

>    and the shape of the data, as appropriate; please include a

Reviewer: Part about estimators and shape of data probably not relevant

   reproducible_ code snippet
   or link to a gist_. If an exception is raised,
   please provide the traceback.

.. _Creating and highlighting code blocks: https://help.github.com/articles/creating-and-highlighting-code-blocks
.. _issues: https://github.com/paris-saclay-cds/ramp-board/issues
.. _pull requests: https://github.com/paris-saclay-cds/ramp-board/pulls
.. _reproducible: https://stackoverflow.com/help/mcve
.. _gist: https://gist.github.com

Documentation
=============

We are glad to accept any sort of documentation: function docstrings,
reStructuredText documents (like this one), tutorials, etc.
reStructuredText documents live in the source code repository under the
doc/ directory.

You can edit the documentation using any text editor and then generate
the HTML output by typing ``make html`` from the doc/ directory.
Alternatively, ``make`` can be used to quickly generate the

> documentation without the example gallery. The resulting HTML files will

Reviewer: I don't think ramp-board has an example gallery, so you can probably remove this part.

be placed in _build/html/ and are viewable in a web browser. See the

> README file in the doc/ directory for more information.

Reviewer: There is no README file in `doc` directory.

> For building the documentation, you will need 

Reviewer: Not sure where the list of required packages for building documentation would be. Maybe [here](https://github.com/paris-saclay-cds/ramp-board/blob/master/ci_tools/circle/build_doc.sh)?

On travis it seems they install all the required packages for ramp board, and then add:
sphinx sphinx_rtd_theme numpydoc graphviz eralchemy sphinx-click. 

Not sure if you need to install ramp board to be able to build the documentation, @glemaitre ?

    - sphinx_, 
    - matplotlib_, and
    - pillow_.

.. _sphinx: http://sphinx-doc.org
.. _matplotlib: https://matplotlib.org
.. _pillow: https://pillow.readthedocs.io

When you are writing documentation, it is important to keep a good
compromise between mathematical and algorithmic details, and give
intuition to the reader on what the algorithm does. It is best to always
start with a small paragraph with a hand-waving explanation of what the
method does to the data and a figure (coming from an example)

> illustrating it.

Reviewer: This part is probably only relevant to scikit learn

.. _contribute:

########################
Develop and contribute
########################

Welcome to the RAMP team. We are always happy to have new RAMP developers.

You can contribute to this code by making a `Pull Request 
<https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-requests>`_ 
on Github_. Please, make sure that your code is coming with unit tests to 
ensure full coverage and continuous integration in the API.

.. _GitHub: https://github.com/paris-saclay-cds/ramp-board/pulls

Install for development
-----------------------
To install RAMP please fallow :ref:`install guideline <install>` making sure 

> that you use the ``make inplace`` option.

Reviewer 2: ```suggestion
that you use the ``make inplace`` option to install RAMP in developer mode.
```

Prepare database engine
-----------------------

> Next, create the database, following the steps as described here:

Reviewer 2: ```suggestion
Testing RAMP require a database, you need to create a database similarly to
```

:ref:`create database <set_database>`

> You created the ``Postgres database cluster`` using the command::

Reviewer 2: ```suggestion
In short, create the ``Postgres database cluster`` using the command::
```

    ~ $ initdb postgres_dbs

> and then you started it with::

Reviewer 2: ```suggestion
and then start it with::
```

    ~ $ pg_ctl -D postgres_dbs -l logfile start

However, RAMP tests do not know which database engine you are using for 
your ramp development and therefore you need to inform RAMP tests about this
To do this, open the ``db_engine.yml`` file located in the ``ramp-board`` directory. 
It should look as follows::

    db_owner: postgres
    db_cluster_name: postgres

> The above settings are used by ``Travis Cl`` which is a continuous integration

Reviewer 2: You might remove the info about travis instead to say something like

```suggestion
where `db_owner` is ... and `db_cluster_name` is ...
```

Reviewer 2: Basically as you mentioned right after :)

service hosted at ``GitHub``. Therefore, you need to change them to your local
settings.

``db_owner`` is the owner of your database. If you don't know who is 
the owner of your database you can find it out by typing in your terminal::
    
    ~ $ psql -l

This command will list all of your databases along with their owners. 

``db_cluster_name`` is the name of your Postgres database cluster (if you used

> commands as above your ``Postgres database cluster`` name is 'postgres_dbs').

Reviewer 2: ```suggestion
commands as above your ``Postgres database cluster`` name is `'postgres_dbs'`).
```

Reviewer: I got this after running `psql -l`:

```
                                 List of databases
      Name      | Owner  | Encoding |   Collate   |    Ctype    | Access privileges 
----------------+--------+----------+-------------+-------------+-------------------
 databoard_test | mrramp | UTF8     | en_US.UTF-8 | en_US.UTF-8 | 
 postgres       | lucy   | UTF8     | en_US.UTF-8 | en_US.UTF-8 | 
 template0      | lucy   | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/lucy          +
                |        |          |             |             | lucy=CTc/lucy
 template1      | lucy   | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/lucy          +
                |        |          |             |             | lucy=CTc/lucy

```
There is no `postgres_dbs`? Did I do something wong?

Test
----

> Each time before running the tests make sure your Postgres database cluster has been

Reviewer 2: ```suggestion
Each time before running the tests make sure your ``Postgres database cluster`` has been
```

started. You can always start it using the command::

>     ~ $ pg_ctl -D postgres_dbs -l logfile start

Reviewer 2: ```suggestion
    ~ $ pg_ctl -D postgres_dbs -l logfile start
    
where `postgres_dbs` is ...
```

> You will also need SMTP server, to run it type::

Reviewer 2: ```suggestion
In addition, you might want to start an SMTP server to run all the tests.
If you don't run the server, some of the tests will be marked as `xfail` 
because they cannot be run. To launch the server, execute the following 
command::`
You will also need SMTP server, to run it type::
```

Author: atm the tests fail if there is no smtp server because there is no way of checking if smtp server is on at the test side

        $ pip install pytest pytest-cov
        $ pytest -vsl .

-  No flake8 warnings (which includes pep8 and pyflakes), check with::

        $ pip install flake8
        $ flake8 path/to/module.py

Filing bugs
===========
We use Github issues to track all bugs and feature requests; feel free to
open an issue if you have found a bug or wish to see a feature implemented.

It is recommended to check that your issue complies with the
following rules before submitting:

-  Verify that your issue is not being currently addressed by other
   issues_ or `pull requests`_.

-  Please ensure all code snippets and error messages are formatted in
   appropriate code blocks.
   See `Creating and highlighting code blocks`_.

-  Please include your operating system type and version number, as well
   as your Python, scikit-learn, numpy, and scipy versions. This information
   can be found by runnning the following code snippet::

    python
    import platform; print(platform.platform())
    import sys; print("Python", sys.version)
    import numpy; print("NumPy", numpy.__version__)
    import scipy; print("SciPy", scipy.__version__)

>     import sklearn; print("Scikit-Learn", sklearn.__version__)

Reviewer 2: Here we just need to know the version of the different ramp packages instead of numpy, scipy, and scikit-learn

Author: @glemaitre Thanks for your comments. I am not sure if I understand this wone. What versions of ramp packages?

   to: indicate you are working on something to avoid duplicated work,
   request broad review of functionality or API, or seek collaborators.
   WIPs often benefit from the inclusion of a `task list`_ in the PR description.

-  All other tests pass when everything is rebuilt from scratch. On
   Unix-like systems, check with (from the toplevel source folder)::

        $ make

-  Documentation and high-coverage tests are necessary for enhancements
   to be accepted.

-  At least one paragraph of narrative documentation with links to
   references in the literature (with PDF links when possible) and
   the example.

.. _coding-guidelines: http://scikit-learn.org/dev/developers/contributing.html#coding-guidelines
.. _task list: https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments

You can also check for common programming errors with the following
tools:

-  Code with good unittest coverage (at least 80%), check with::

        $ pip install pytest pytest-cov
        $ pytest -vsl .

-  No pyflakes warnings, check with::

        $ pip install pyflakes
        $ pyflakes path/to/module.py

-  No PEP8 warnings, check with::

        $ pip install pep8

>         $ pep8 path/to/module.py

Reviewer 3: I would recomment flake8 that does both. And I would not recomment autopep8 as it can do horrors.

it's just a nitpick. Besides looks great !
